File: core/js/state.mjs
/* state.mjs — ShopState (mirror of core/state.py).
 *
 * The generic shop-side context the cost model reads: a clock, finite
 * inventory, slot capacity, perishable batches, an expiring-option set, and an
 * expected-demand table.  Nothing vertical-specific.
 *
 * Representation choices to mirror Python semantics exactly:
 *   inventory / expected_demand -> plain objects keyed by option id (string)
 *   capacity                    -> Map keyed by slot_ticks (int); may hold
 *                                  -Infinity (a slot the adapter force-drops)
 *   expiring                    -> Set of option ids
 */

export class Batch {
  constructor(servings, expires_tick, key = "") {
    this.servings = servings;
    this.expires_tick = expires_tick;
    this.key = key;
  }
}

export class ShopState {
  constructor({
    tick = 0,
    inventory = {},
    capacity = new Map(),
    batches = [],
    expiring = new Set(),
    expected_demand = {},
    extra = {},
  } = {}) {
    this.tick = tick;
    this.inventory = inventory;
    this.capacity = capacity instanceof Map ? capacity : new Map(Object.entries(capacity).map(([k, v]) => [Number(k), v]));
    this.batches = batches;
    this.expiring = expiring instanceof Set ? expiring : new Set(expiring);
    this.expected_demand = expected_demand;
    this.extra = extra;
  }

  stock(optionId) {
    return Object.prototype.hasOwnProperty.call(this.inventory, optionId)
      ? this.inventory[optionId]
      : 0.0;
  }
}

if (typeof globalThis !== "undefined") {
  globalThis.SNHP_state = { Batch, ShopState };
}

File: core/state.py
"""ShopState — the generic shop-side context the cost model reads.

boba.world.ShopState (clock + FIFO queue + tapioca batches) and
vend.core.MachineState (clock + per-SKU lots + par stock) are two concrete
instances of the same idea: a clock, some finite inventory, some capacity,
and some perishable batches. The core cost components (core/cost.py) read
only the generic fields below, so one state type feeds both worlds.

Nothing here is vertical-specific: an adapter (Phase 2/3) projects boba's or
vend's live state onto these fields before quoting.
"""
from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class Batch:
    """A perishable lot: `servings` left, dies at `expires_tick`."""
    servings: int
    expires_tick: int
    key: str = ""            # which option/sku this batch belongs to


@dataclass
class ShopState:
    tick: int = 0
    # option_id -> units on hand (finite stock; read by scarcity_shadow)
    inventory: dict[str, float] = field(default_factory=dict)
    # slot -> units the slot can still absorb (read by fulfillment guards)
    capacity: dict[int, float] = field(default_factory=dict)
    # perishable lots (read by salvage_on_expiry / expiry accounting)
    batches: list[Batch] = field(default_factory=list)
    # option_ids currently priced at salvage (batch expiring in excess).
    # An adapter sets this from the vertical's clearance rule — boba's
    # pearls_expiring_excess, vend's days_to_expiry<=0.
    expiring: set[str] = field(default_factory=set)
    # option_id -> expected rest-of-horizon list-price demand (read by
    # scarcity_shadow: units in EXCESS of this are cheap to move; the rest
    # would have sold at list, so discounting them displaces that sale).
    expected_demand: dict[str, float] = field(default_factory=dict)
    # escape hatch for adapter-specific facts a bespoke cost fn might read
    extra: dict = field(default_factory=dict)

    def stock(self, option_id: str) -> float:
        return self.inventory.get(option_id, 0.0)
